ios/engine/KMEI/KeymanEngine/Classes/Resource Management/ResourceDownloadManager.swift

```diff
@@ -66,86 +66,56 @@ public class ResourceDownloadManager {
+  where FullID.Resource.Package: TypedKeymanPackage<FullID.Resource> {
+    // Note:  in this case, someone knows the "full ID" of the resource already, but NOT its location.
+    //        We can use the package-version query to attempt a lookup for a .kmp location
+    //        for download.
+    let packageKey = KeymanPackage.Key(id: fullID.id, type: fullID.type)
+    Queries.PackageVersion.fetch(for: [packageKey], withSession: session) { result, error in
+      guard let result = result, error == nil else {
+        log.info("Error occurred requesting location for \(fullID.description)")
+        self.resourceDownloadFailed(forFullID: fullID, with: error ?? .noData)
+        completionBlock?(nil, error ?? .noData)
+        return
       }
-      
-      downloader.queue(.simpleBatch(dlBatch))
-      return dlBatch
-    }
-    return nil
-  }
-  
-  private func buildKeyboardDownloadBatch(for keyboard: InstallableKeyboard,
-                                          withFilename filename: String,
-                                          asActivity activity: DownloadActivityType,
-                                          withOptions options: Options,
-                                          startBlock: (() -> Void)? = nil,
-                                          completionBlock: CompletionHandler<InstallableKeyboard>? = nil) -> DownloadBatch<InstallableKeyboard.FullID>? {
-    let keyboardURL = options.keyboardBaseURL.appendingPathComponent(filename)
-    let fontURLs = Array(Set(keyboardFontURLs(forFont: keyboard.font, options: options) +
-                             keyboardFontURLs(forFont: keyboard.oskFont, options: options)))
 
-    do {
-      try FileManager.default.createDirectory(at: Storage.active.resourceDir(for: keyboard)!,
-                                              withIntermediateDirectories: true)
-    } catch {
-      log.error("Could not create dir for download: \(error)")
-      return nil
-    }
+      guard case let .success(data) = result.entryFor(packageKey) else {
+        if case let .failure(errorEntry) = result.entryFor(packageKey) {
+          if let errorEntry = errorEntry {
+            log.info("Query reported error: \(String(describing: errorEntry.error))")
+          }
+        }
+        self.resourceDownloadFailed(forFullID: fullID, with: Queries.ResultError.unqueried)
+        completionBlock?(nil, Queries.ResultError.unqueried)
+        return
+      }
 
-    var request = HTTPDownloadRequest(url: keyboardURL, userInfo: [:])
-    request.destinationFile = Storage.active.cloudKeyboardURL(forID: keyboard.id).path
-    request.tag = 0
+      // Perform common 'can download' check.  We need positive reachability and no prior download queue.
+      guard self.downloader.state == .clear else {
+        let err = self.downloader.state.error ??
+          NSError(domain: "Keyman", code: 0, userInfo: [NSLocalizedDescriptionKey: "Already busy downloading something"])
```

Comment:
I don't understand why we can only download one thing at a time?

Reply:
It's something that was in place before I ever began working on the code base.  I think it's because downloading was always tied to UI code and the involvement of notifications in finalizing downloads?

That latter point is something that got nixed this phase, so there's a chance we could safely drop it.